kvserver: Reduce replica mutex contention Phase 2

iskettaneh opened the pull request:
The first 2 commits are based on https://github.com/cockroachdb/cockroach/pull/146085

This PR reduces the replica mutex contention for both checkExecutionCanProceedBeforeStorageSnapshot and checkExecutionCanProceedAfterStorageSnapshot by making each of the two methods only briefly take the mutex and release it after it copies everything it needs.

Running this microbenchmark:

```
rm -f /tmp/*.pb.gz && env GODEBUG=runtimecontentionstacks=1 go test ./pkg/sql/tests -run - -bench BenchmarkParallelSysbench/SQL/3node/oltp_read_write -v -test.benchtime=3000x -test.outputdir=/tmp -test.mutexprofile=mutex.pb.gz -test.mutexprofilefraction=100 -test.timeout 25s 2>&1
```

shows that the ReplicaMutex contention goes from `786.9ms` to `614.3ms`.

Before:
<img width="1728" alt="Screenshot 2025-05-06 at 6 04 48 PM" src="https://github.com/user-attachments/assets/1438d395-5edd-4370-a683-dceaee808cee" />

After:
<img width="1725" alt="Screenshot 2025-05-06 at 6 04 35 PM" src="https://github.com/user-attachments/assets/85aa2c19-ac14-48fa-91a0-ad45b2979fe4" />


Running benchdiff command:
```
rm -f /tmp/*.pb.gz && env GODEBUG=runtimecontentionstacks=1 benchdiff -b --old 6cbfc38325b68a2547165ef194699c781b15f476 -c 20 ./pkg/sql/tests -r BenchmarkParallelSysbench/SQL/3node/oltp_read_write --mutexprofile -d=3000x
```

Shows the resutls:

```
name                                           old time/op    new time/op    delta
ParallelSysbench/SQL/3node/oltp_read_write-24    1.39ms ± 4%    1.39ms ± 3%   ~     (p=0.912 n=10+10)

name                                           old errs/op    new errs/op    delta
ParallelSysbench/SQL/3node/oltp_read_write-24      0.01 ±57%      0.01 ±27%   ~     (p=0.092 n=10+10)

name                                           old alloc/op   new alloc/op   delta
ParallelSysbench/SQL/3node/oltp_read_write-24    2.07MB ± 1%    2.07MB ± 1%   ~     (p=0.912 n=10+10)

name                                           old allocs/op  new allocs/op  delta
ParallelSysbench/SQL/3node/oltp_read_write-24     8.29k ± 1%     8.30k ± 1%   ~     (p=0.684 n=10+10)
```


References: #140235

Release note: None

iskettaneh committed:
kvserver: Make sure that (*Replica).shMu.state.Lease is immutable

This commit ensures that (*Replica).shMu.state.Lease is immutable. This
is going to be useful when we start reducing the Replica mutex
contention by reading the state.Lease early, and then releasing the
replica lease.

References: #140235

Release note: None

iskettaneh committed:
kvserver: reduce r.mu contention for checkExecutionCanProceedBeforeStorageSnapshot

This commit makes checkExecutionCanProceedBeforeStorageSnapshot to take
the replica read mutex only for a brief time to take all the fields it
needs. This reduces the replica mutex contention.

References: #140235

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146131)
<!-- Reviewable:end -->

tbg commented on pkg/kv/kvserver/helpers_test.go:
```diff
@@ -708,6 +708,13 @@ func (r *Replica) SupportFromEnabled() bool {
 	return (*replicaRLockedStoreLiveness)(r).SupportFromEnabled()
 }
 
+// SupportFromEnabledWithoutRLock exports
```
Could you clean this up while you're here? `SupportFromEnabled` should be `SupportFromEnabledRLocked()`, and your new method should be `SupportFromEnabled`.

Disregard if this is gone in the end.

iskettaneh replied:
My understanding is that `SupportFromEnabled` is an interface method. Nothing really says that replica mutex should be taken. To fix this issue, the team created a new type "replicaRLockedStoreLiveness" that implements this interface.

Anyways, now updated the code to have 2 versions:

1) (*replicaRLockedStoreLiveness) SupportFromEnabled() -> Requires RLock 
2) (*Replica) SupportFromEnabled(desc *roachpb.RangeDescriptor) -> Doesn't require the RLock
<!-- Reviewable comment -OPkurroDLbXGyeFsi2q:b-ogmxhy -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/replica.go:
```diff
@@ -1727,6 +1727,19 @@ func (r *Replica) getReplicaDescriptorRLocked() (roachpb.ReplicaDescriptor, erro
 	return roachpb.ReplicaDescriptor{}, kvpb.NewRangeNotFoundError(r.RangeID, r.store.StoreID())
 }
 
+// getReplicaDescriptor is similar to getReplicaDescriptorRLocked but doesn't
```
please implement `getReplicaDescriptorRLocked` by calling `getReplicaDescriptor`. (Disregard if the RLocked is gone in the end).

iskettaneh replied:
Done.
<!-- Reviewable comment -OPkp3SGBGxZzDdDkyqF:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/replica.go:
```diff
@@ -2123,21 +2133,38 @@ func (r *Replica) checkExecutionCanProceedBeforeStorageSnapshot(
+	// from the replica while we hold the RLock and then release it.
+	desc := r.descRLocked()
+	mergeInProgress := r.mergeInProgressRLocked()
+	cachedClosedTimestampPolicy := *r.cachedClosedTimestampPolicy.Load()
```
Not technically new, but let's think about it: why do this under the mutex? Is this field only updated under `r.mu` and we use this to synchronize properly with updates to the policy? Looking at `RefreshPolicy`, there is no `r.mu` lock, and correspondingly we should be able to move this load outside the critical section.

@wenyihu6 might have thoughts on whether there are any invariants to be mindful of here; please update the comment on the field if we learn anything.

https://github.com/cockroachdb/cockroach/blob/561a43d20ce870dbe0ad12fc3e51e3bee49c60e7/pkg/kv/kvserver/replica.go#L1407-L1458

iskettaneh replied:
Agreed. I actually had it out for the most time during development. I just pushed it last thing before uploading the PR for no good reason.
<!-- Reviewable comment -OPlHUWN7CKxJO4jV0Wj:bd7926q -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/replica.go:
```diff
@@ -2147,14 +2174,14 @@ func (r *Replica) checkExecutionCanProceedBeforeStorageSnapshot(
 		// is false (e.g. RequestLeaseRequests) will not wait for a pending
 		// merge before executing and, as such, can execute while a range is in
 		// a merge's critical phase (i.e. while the RHS of the merge is
 		// subsumed).
```
Add the following comments here if they make sense to you:

// Note that we are not relying on r.mu synchronization here. On the
// RHS leaseholder that originally sees the Subsume, we hold latches
// across all keys[1a] while calling WatchForMerge[1b].
// When a new leaseholder steps up, it installs the
// merge watcher channel in leasePostApply, before the lease can be
// used to serve requests.
//
// [1a]: see batcheval.declareKeysSubsume.
// [1b]: see batcheval.Subsume.
// [2]: see leasePostApply.

I had to convince myself that this wasn't relying on being in a critical section somehow.

iskettaneh replied:
Done.
<!-- Reviewable comment -OPksRSA1c53Vt7JHwRo:b-896fix -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/replica_range_lease.go:
```diff
@@ -1285,7 +1300,8 @@ func (r *Replica) redirectOnOrAcquireLeaseForRequest(
 
 			now := r.store.Clock().NowAsClockTimestamp()
-			status := r.leaseStatusForRequestRLocked(ctx, now, reqTS)
+			status := r.leaseStatusForRequest(ctx, now, reqTS, r.mu.minLeaseProposedTS,
```
We're leaving things in a somewhat awkward in-between place with all of these methods that sit on `*Replica` but don't assume a lock taken. This is fine for now, just letting you know that I'm mulling this over.

iskettaneh replied:
Ack
<!-- Reviewable comment -OPlHfTtEmvQu-lze-18:b2nhb2c -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/replica_store_liveness.go:
Yes, please address this in this PR. If this is somehow difficult, let's talk!

Btw I'm always going to call out the `WithoutRLock` pattern, as our established style is that the method that needs the lock needs to have the suffix, and the method that doesn't need the lock (or acquires it itself) doesn't have a suffix. (The latter is arguably not ideal, since a method that acquires a lock can't be called when the lock is held; we currently rely on ... well, knowing what's what, but this gets harder now that we're introducing new flavors of non-locked methods on Replica, though they're hopefully not going to stick for long).

iskettaneh replied:
I agree with you that I also don't like the WithoutRLock suffix.  I gave this another thought. Please see my response on your comment below.
<!-- Reviewable comment -OPlHFEJ3iBNzhFgVkHt:bmerh6a -->
<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
Didn't review the last commit (kvserver: reduce r.mu contention for checkExecutionCanProceedAfterSto…) yet.

Thanks for all the refactoring! This looks roughly like I was expecting. Some comments, none super major, but still should be worked out in the context of the current PR.

iskettaneh committed:
kvserver: reduce r.mu contention for checkExecutionCanProceedAfterStorageSnapshot

This commit makes checkExecutionCanProceedAfterStorageSnapshot to take
the replica read mutex only for a brief time to take all the fields it
needs. This reduces the replica mutex contention.

References: #140235

Release note: None

iskettaneh reviewed (commented):
Thanks for the review. I removed the last commit from this PR to make this PR a bit easier to review

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146131#-:-OPko0Yf86z7nNwZoI8j:b-an8ry6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)



<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
Change in plans. I re-added the last commit since it was small, and kinda relevant to the PR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146131#-:-OPlLi_t0Vmgpt9LheqV:b38gemh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)



<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Looks good, thanks! Can you add benchdiff results before you merge?

iskettaneh commented:
I added the benchdiff results. Thanks for the review!

bors r+
